xds/src/main/java/com/linecorp/centraldogma/xds/k8s/v1/XdsKubernetesService.java

```diff
@@ -96,43 +102,57 @@ public void createServiceEndpointWatcher(CreateServiceEndpointWatcherRequest req
-        final ServiceRequestContext ctx = ServiceRequestContext.current();
-
-        // Use a schedule to time out the watcher creation until we implement a custom implementation.
-        final ScheduledFuture<?> scheduledFuture = ctx.eventLoop().schedule(() -> {
-            if (!completed.compareAndSet(false, true)) {
-                return;
-            }
-            kubernetesEndpointGroup.closeAsync();
-            responseObserver.onError(
-                    Status.INTERNAL.withDescription(
-                            "Failed to retrieve k8s endpoints within 5 seconds. watcherName: " +
-                            watcher.getName()).asRuntimeException());
-        }, 5, TimeUnit.SECONDS);
-
-        whenReady.handle((endpoints, cause) -> {
-            if (!completed.compareAndSet(false, true)) {
-                return null;
-            }
-            scheduledFuture.cancel(false);
-            kubernetesEndpointGroup.closeAsync();
+        final ContextAwareBlockingTaskExecutor taskExecutor =
+                ServiceRequestContext.current().blockingTaskExecutor();
+        final CompletableFuture<KubernetesEndpointGroup> future =
+                createKubernetesEndpointGroup(watcher, xdsResourceManager.xdsProject().metaRepo(),
+                                              taskExecutor);
+        future.handle((kubernetesEndpointGroup, cause) -> {
             if (cause != null) {
-                // Specific types.
-                responseObserver.onError(Status.INTERNAL.withCause(cause).asRuntimeException());
+                cause = Exceptions.peel(cause);
+                if (cause instanceof IllegalArgumentException || cause instanceof EntryNotFoundException) {
+                    responseObserver.onError(Status.INVALID_ARGUMENT.withCause(cause).asRuntimeException());
+                } else {
+                    responseObserver.onError(Status.INTERNAL.withCause(cause).asRuntimeException());
+                }
                 return null;
             }
-            logger.debug("Successfully retrieved k8s endpoints: {}", endpoints);
-            onSuccess.run();
+            final AtomicBoolean completed = new AtomicBoolean();
+            final CompletableFuture<List<Endpoint>> whenReady = kubernetesEndpointGroup.whenReady();
+
+            // Use a schedule to time out the watcher creation until we implement a custom implementation.
+            final ScheduledFuture<?> scheduledFuture = taskExecutor.schedule(() -> {
+                if (!completed.compareAndSet(false, true)) {
+                    return;
+                }
+                kubernetesEndpointGroup.closeAsync();
+                responseObserver.onError(
+                        Status.INTERNAL.withDescription(
+                                "Failed to retrieve k8s endpoints within 5 seconds. watcherName: " +
+                                watcher.getName()).asRuntimeException());
+            }, 5, TimeUnit.SECONDS);
```

Comment:
Q) 5 seconds seems short compared to request timeouts. Why did you set 5 seconds as the deadline?

Reply:
I chose a value of less than 10 seconds, which is the request timeout.
This API has two phases: calling k8s and committing.
I estimated that the API would need at least 5 seconds for the committing phase so I allocated 5 seconds for the k8s.
I could potentially increase the default timeout and assign the larger value but I felt it wasn't necessary at this point. 
Do you have any recommendations?